pkg/docker/config/config.go

```diff
@@ -552,6 +575,14 @@ func readJSONFile(path string, legacyFormat bool) (dockerConfigFile, error) {
 		return dockerConfigFile{}, err
 	}
 
+	// When creating a file for the first time, it may be empty at this
+	// point as it was created by the write lock. Hence, return early here
```

Comment:
That’s surprising to me (and incompatible with older versions of consumers, though that’s a small concern — they could easily race with `WriteFile` and see an empty file, and fail, for that reason).

See elsewhere about sharing the lock with JSON.

Reply:
I think it's a valid concern but I do not have a good/definitive answer/idea.  We could use one central (similar to the short-names lock) _but_ old client would still compete with newer ones.

Reply:
I was thinking about using a separate `auth.json.lock` file; then the concerns about the lock file object trying to own file contents would go away, and so would this situation, AFAICS.

Reply:
The underlying `fcntl` lock is unlocked when _any_ file descriptor to that file is closed, isn’t it?

If so, `ioutil.ReadFile` of the JSON always releases the lock, making the `modifyJSON` lock ineffective — and we do need to use a different path.

Reply:
> I was thinking about using a separate `auth.json.lock` file; then the concerns about the lock file object trying to own file contents would go away, and so would this situation, AFAICS.

I was considering that but didn't like the user experience.  I would find it surprising/annoying if I do a `--authfile file` and would find a `file.lock` afterwards.

> The underlying `fcntl` lock is unlocked when _any_ file descriptor to that file is closed, isn’t it?

Holy frog, that is true and buried in the man pages. That indeed is a K.O. for this approach - we could do ref-counting on the path but that seems a bit too hacky for my taste.

Reply:
Just to be explicit,`pkg/lockfile` is already using a global hash table and refcounting the objects to work correctly with those locks, so that works for exclusively-used lockfiles (and as long as c/storage doesn’t get a separate `/v2` with a separate hash table); but it breaks if there is any _other_ user of that file, like here.

Reply:
… the point about `file.lock` files on user-specified `--authfile` paths (outside of dot files, in user-maintained project repositories and the like) being annoying is quite valid, though; we might not even necessarily be able to create a lockfile (if the `auth.json` is in a read-only configuration directory).

flock(2) might be an option, but it is distinctly less portable (and the Linux man pages suggest that some implementations implement it on top of the `fcntl` locks. I’m sure there are good write-ups about the various locking mechanisms and their trade-offs, I haven’t now taken the time to find and read one.

Reply:
A possible _partial_ alternative: use `c/storage/pkg/ioutils.AtomicWriteFile`. That could still lose data with concurrent writers, but at least it wouldn’t break readers concurrent with a single writer, who can now see empty files or partial contents (= invalid JSON).

Reply:
Seems like we're between a rock and hard place.  I will close this PR as I ran out of time and really do not have a good idea yet.

Reply:
Let’s at least do the `AtomicWriteFile` part, that should be safe and make _some_ difference: https://github.com/containers/image/pull/1515 .